src/posthog-core.ts

```diff
@@ -1074,6 +1105,8 @@ export class PostHog {
             this.register({ distinct_id: new_distinct_id })
         }
 
+        // historically we checked for a missing device id *or* distinct_id == device_id as a proxy for detecting anonymous users
+        // we now have the explicit anonymous user state but we still check for a missing device id to maintain backwards compatibility
         const deviceIdMarksForIdentify = !this.get_property('$device_id')
         const isKnownAnonymous = this.persistence.get_user_state() === 'anonymous'
 
```

Comment:
Can we add a comment below sth like:

if the distinct ID changes, then send an `$identify` event, if it didn't send a `$set` or `$set_once` event.

Maybe add sth like: One cannot identify an already identified user into another identified user. We have this safeguard to avoid merging user together who shouldn't be (e.g. identify('user') -> identify('staffUser') -> identify('user')). This is enforced in the ingestion backend already, but to avoid bad event ingestion ordering causing havoc we enforce this here too.

Actually this is not nice for the user if we go into the else branch we send 2 events instead of 1 iiuc, which increases the event count unnecessarily.